microprofile/server/src/main/java/io/helidon/microprofile/server/ServerCdiExtension.java

```diff
@@ -385,6 +390,33 @@ private void addApplication(JaxRsCdiExtension jaxRs, JaxRsApplication applicatio
         }
     }
 
+    private final Map<Service, AnnotatedMember<?>> producedServices = new HashMap<>();
+
+    private void registerServiceProducers(@Observes ProcessProducer<?, Service> pp) {
+        Producer<Service> producer = pp.getProducer();
+        AnnotatedMember<?> annotatedMember = pp.getAnnotatedMember();
+
+        pp.setProducer(new Producer<>() {
+            @Override
+            public Service produce(final CreationalContext<Service> ctx) {
+                Service service = producer.produce(ctx);
+                producedServices.put(service, annotatedMember);
```

Comment:
Conceivably this could be called on any thread (whenever someone injects a `Service`, wherever they do so), so `producedServices` will need to be thread safe.  Also I don't see any matching deletion?

Reply:
This also relies on `Service` equality to be unchanging.  I don't think `Service` has an `equals`/`hashCode` contract, does it?

Reply:
Thx! Fixed with synchronized IdentityHashMap